native/spark-expr/src/structs.rs

```diff
@@ -125,3 +125,103 @@ impl PartialEq<dyn Any> for CreateNamedStruct {
             .unwrap_or(false)
     }
 }
+
+#[derive(Debug, Hash)]
+pub struct GetStructField {
+    child: Arc<dyn PhysicalExpr>,
+    ordinal: usize,
+}
+
+impl GetStructField {
+    pub fn new(child: Arc<dyn PhysicalExpr>, ordinal: usize) -> Self {
+        Self { child, ordinal }
+    }
+
+    fn child_field(&self, input_schema: &Schema) -> DataFusionResult<Arc<Field>> {
+        match self.child.data_type(input_schema)? {
+            DataType::Struct(fields) => Ok(fields[self.ordinal].clone()),
+            data_type => Err(DataFusionError::Plan(format!(
+                "Expect struct field, got {:?}",
+                data_type
+            ))),
+        }
+    }
+}
+
+impl PhysicalExpr for GetStructField {
```

Comment:
Don't know enough about DataFusion to really know what the difference is. Just on the Spark side, UDFs are usually slightly less performant, so if you don't have to do a UDF you're usually better off. DataFusion does have a [get_field](https://github.com/apache/datafusion/blob/main/datafusion/functions/src/core/getfield.rs) ScalarUDF already it looks like, but that's by name and not index, and there seems like a lot more ceremony about checking all the input types, vs the PhysicalExpr is more tailored to what we already know about the input data.

Reply:
I agree that we should start implementing functions as ScalarUDFImpl instead of PhysicalExpr. I think it would be fine to convert this one as a follow on PR.

Reply:
What's the tl;dr on the benefits of that? Easier to use outside of the Comet/Spark use case?

Reply:
Yes, I think it makes it easier for DataFusion users to switch between different function implementations and can be used from the logical plan as well as from the physical plan

Reply:
In this particular case `GetStructField` might not even make sense as a UDF because it's created by the Spark analyzer from an `ExtractValue` expression, which resolves the name to an ordinal in the struct. It'd be odd to use `GetStructField` directly, since you would normally want a nested column by name, and not by index. I think the existing `get_field` UDF in DataFusion covers that use case, but using this PhysicalExpr on the Spark side seems like it makes more sense instead of having to convert the ordinal back into a field name to then use `get_field`